source/utils/numeric.cpp

```diff
@@ -0,0 +1,64 @@
+// Copyright (c) 2024 xlnt-community
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE
+//
+// @license: http://www.opensource.org/licenses/mit-license.php
+// @author: see AUTHORS file
+
+#include <xlnt/utils/numeric.hpp>
+#include <detail/locale.hpp>
+#include <detail/parsers.hpp>
+
+#include <iomanip>
+
+constexpr auto serialization_digits_full_precision = 15;
+constexpr auto serialization_digits_short = 6;
+
+std::string xlnt::detail::number_serialiser::serialise(double d) const
+{
+    std::ostringstream ss;
```

Comment:
`ostringstream` may be slower on most systems compared to `snprintf`. Maybe benchmarking may be relevant as this function may be called a lot.

Reply:
Running the `spreadsheet-load` benchmark (which also contains a write benchmark) reveals a speed decrease of ~20% for writing to an xlsx file on my machine.

Reply:
Thanks for benchmarking! This is indeed not good. I preferred using `std::ostringstream` for formatting, as it already outputs the correct format right away and needs no further replacements. However, you proved once again that streams in C++ are such a mess - the reason why I try to avoid them if possible.

To be honest, I'm neither happy with the old solution, nor with my new one. Both solutions are bad, which is why `number_formatter::fill_placeholders` requires the following workaround:
```cpp
        auto s = serialiser_.serialise_short(number);
        while (s.size() > 1 && s.back() == '0')
        {
            s.pop_back();
        }
        if (s.back() == '.')
        {
            s.pop_back();
        }
        return s;
```

I tried to fix it, however, `std::setprecision` does not work with `std::defaultfloat` the same way it works with `std::fixed` (with `std::defaultfloat`, setting precision to 6 does not mean 6 digits after the decimal point, but a total number of 6 digits (e.g. 4 before decimal and 2 afterwards, or other combinations)), which is why I switched back to `std::fixed` with the same workaround as before. So neither the old nor the new solutions are ideal, but the old one is faster.

What about adding the `fmt` library, which should fix our issues without further workarounds? It should also improve performance according to many tests, [including this one](https://github.com/fmtlib/fmt?tab=readme-ov-file#speed-tests).

Reply:
You may include the `fmt` library if you want, but it may not be strictly necessary if we use `std::from_chars` and `std::to_chars` if c++17 support is enabled.

Reply:
While I do agree, this would also mean that compiling in C++17 mode (or newer, in the future) would produce a different output than compiling in C++11 or C++14 modes. In other words, we would need the workarounds above in C++11 and C++14 modes, but not in C++17 or newer. This would be inconsistent and error-prone, so we would need to be a lot more careful in such cases. Thankfully, we have unit tests, but it can always happen that we forget to test something :sweat_smile: I would personally prefer to add the `fmt` library, as this would give us the best performance in all C++ versions and would ensure consistency. If you agree, I could try to add it and see if we are happy with it :slightly_smiling_face:

Reply:
I don't think it is that bad to have two implementations in this case, but feel free to include the `fmt` library if you like. Just don't forget to insert the fmt license to `license.md`.

Reply:
I don't mean that it's bad to have two implementations, but it's not great to have two implementations that behave differently. Because in that case we also need to have the same version checks all over the place where workarounds are necessary, like this one:
```cpp
        auto s = serialiser_.serialise_short(number);
        while (s.size() > 1 && s.back() == '0')
        {
            s.pop_back();
        }
        if (s.back() == '.')
        {
            s.pop_back();
        }
        return s;
```

With different implementations, we would need to have different implementations for such code as well. This is what I mean by being error-prone, especially if we forget to write unit tests in some cases.

Reply:
I definitely agree that if we have two implementations for the same function, they should behave (logically) identical.